### PR TITLE
feat(forge): watch mode for forge build

### LINKED ISSUE #813: Add support for --watch mode when running tests
### Component

Forge

### Describe the feature you would like

Enable tests to be run in `--watch` mode, similar to how Jest works in TS/JS projects. 

When running watch mode, should re-run relevant tests whenever any files change in project src.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Fixes #813

## Solution
this adds a `forge build -w|--watch` flag which enters watch mode that runs forge build on changes.

`forge test -w` is slightly more complicated because ideally, we want to adjust the match argument. I think I've figured out how to implement this and will try to add this.

### Update
`forge test --watch` added, currently it's configured to only run those files/file that was changed, see `forge test --match-path`
This will benefit once we can limit the files we need to compile based on the tests we need to run.

The watchexec API is a bit clunky when it comes to reconfiguring the runtime, but it appears to work.

### CODE REVIEW

File: cli/src/cmd/watch.rs [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+        if signals.contains(&MainSignal::Terminate) {
+            action.outcome(Outcome::both(Outcome::Stop, Outcome::Exit));
+            return fut
+        }
+
+        if signals.contains(&MainSignal::Interrupt) {
+            action.outcome(Outcome::both(Outcome::Stop, Outcome::Exit));
+            return fut
+        }
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:233]: nbd but can combine these 2 in one conditional

File: cli/src/cmd/watch.rs [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+                    Some(ProcessEnd::Exception(ex)) => {
+                        (format!("Command ended by exception {:#x}", ex), true)
+                    }
+                    Some(ProcessEnd::Success) => ("Command was successful".to_string(), false),
+                    None => ("Command completed".to_string(), false),
+                };
+
+                if printit {
+                    eprintln!("[[{}]]", msg);
+                }
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:268]: Maybe no need to print these? Could perhaps but them behind `tracing::trace`? Seems to clutter command output a bit

File: cli/src/cmd/watch.rs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+
+                action.outcome(Outcome::DoNothing);
+                return fut
+            }
+        }
+
+        f(OnActionState {
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:275]: so in our cases, this is `on_test` for `test --watch` and a no-op for build --watch

File: cli/src/cmd/watch.rs [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+                Outcome::both(Outcome::Stop, Outcome::both(Outcome::Clear, Outcome::Start))
+            }
+            (false, "restart") => Outcome::both(Outcome::Stop, Outcome::Start),
+            _ => Outcome::DoNothing,
+        };
+
+        let when_idle =
+            if clear { Outcome::both(Outcome::Clear, Outcome::Start) } else { Outcome::Start };
+
+        action.outcome(Outcome::if_running(when_running, when_idle));
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:297]: shockingly low level API

File: cli/src/cmd/watch.rs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        multiple_occurrences = false
+    )]
+    pub watch: Option<Vec<PathBuf>>,
+}
+
+/// Returns the Initialisation configuration for [`Watchexec`].
+pub fn init() -> eyre::Result<InitConfig> {
+    let mut config = InitConfig::default();
+    config.on_error(SyncFnHandler::from(|data| -> std::result::Result<(), Infallible> {
+        eprintln!("[[{:?}]]", data);
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:176]: same, maybe tracing::trace? tracing::error will always be logged which isn't ideal either

File: cli/src/cmd/watch.rs [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    let re_str = format!("({})", sol_files.join("|"));
+    if let Ok(re) = Regex::from_str(&re_str) {
+        let mut new_cmd = cmd.clone();
+        new_cmd.push("--match-path".to_string());
+        new_cmd.push(re.to_string());
+        // reconfigure the executor with a new runtime
+        let mut config = runtime.clone();
+        config.command(new_cmd);
+        // re-register the action
+        on_action(args.clone(), config, wx, cmd, has_conflicting_pattern_args, on_test);
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:105]: interesting that this doesn't seem to infinitely recurse? on_action -> on_test -> on_action

File: cli/src/cmd/watch.rs [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+        // --match-path requires 1 argument
+        cmd.drain(pos..=(pos + 1));
+    }
+
+    // append `--match-path` regex
+    let re_str = format!("({})", sol_files.join("|"));
+    if let Ok(re) = Regex::from_str(&re_str) {
+        let mut new_cmd = cmd.clone();
+        new_cmd.push("--match-path".to_string());
+        new_cmd.push(re.to_string());
```
Comment (gakonst) [inline cli/src/cmd/watch.rs:100]: ok so even if the user has specified a match-path argument, it gets overwritten with the files that were changed in this run. i guess there's an edge case here, where if I have supplied `--match-path A.sol` and the file that changed is B.sol, tests for B.sol end up getting run, when I only wanted A.sol? same issue with inheritance etc., seems tricky to cover extensively, maybe we should document it as a known limitation?

**Review by gakonst — APPROVED** [review]
nice work - lots of pain to integrate though

File: cli/src/cmd/watch.rs (hunk 6 above)
Comment (mattsse) [reply cli/src/cmd/watch.rs:105]: yeah this was surprising and is even wrong in the watchexec example because there's no shared state in the RuntimeConfig that can be mutated if it's cloned before moved to the `Watchexec::new`
and `RuntimeConfig::on_action` does this:

```rust
self.action.action_handler = HandlerLock::new(Box::new(handler));
```

so in order to dynamically reconfigure the `Watchexec` we need this approach which kinda looks like a footgun

File: cli/src/cmd/watch.rs (hunk 7 above)
Comment (mattsse) [reply cli/src/cmd/watch.rs:100]: yeah that could be unexpected, so I assume the correct way would be to leave the command unchanged if it was initially run with a --match-path.

File: cli/src/cmd/watch.rs (hunk 3 above)
Comment (mattsse) [reply cli/src/cmd/watch.rs:275]: exactly, we don't need to modify the command for forge build but want to reconfigure forge test to only run tests in files that were changed, which requires `Watchexec::reconfigure` which is a bit of a pain.

File: cli/src/cmd/watch.rs (hunk 4 above)
Comment (mattsse) [reply cli/src/cmd/watch.rs:297]: sigh

File: cli/src/cmd/watch.rs (hunk 5 above)
Comment (mattsse) [reply cli/src/cmd/watch.rs:176]: ah yeah these include all watchexec related errors as well, lets trace them instead
